Firebase/InstanceID/FIRInstanceID.m

```diff
@@ -977,7 +922,23 @@ - (void)defaultTokenWithHandler:(FIRInstanceIDTokenHandler)handler {
                                 @"Successfully fetched default token.");
       }
       // Post the required notifications if somebody is waiting.
-      if (shouldNotifyHandler && handler) {
+      if (shouldNotifyHandler) {
+        FIRInstanceIDLoggerDebug(kFIRInstanceIDMessageCodeInstanceID008, @"Got default token %@",
+                                 token);
+        NSString *previousFCMToken = self.defaultFCMToken;
+        self.defaultFCMToken = token;
+
+        // Only notify of token refresh if we have a new valid token that's different than before
+        if (self.defaultFCMToken.length &&
+            ![self.defaultFCMToken isEqualToString:previousFCMToken]) {
+          NSNotification *tokenRefreshNotification =
+              [NSNotification notificationWithName:kFIRInstanceIDTokenRefreshNotification
+                                            object:[self.defaultFCMToken copy]];
+          [[NSNotificationQueue defaultQueue] enqueueNotification:tokenRefreshNotification
+                                                     postingStyle:NSPostASAP];
+        }
+      }
+      if (handler) {
```

Comment:
Looks like before the handler was called only when `shouldNotifyHandler == YES` but now it is called regardless the flag. Is it an intentional change?

Reply:
I was actually starting with making the class more readable, then I realize it actually resolves one of the pending issue.

Yes a handler passing in should always be triggered. The notify flag was for whether to send local notification to notify users the token has changed.

Reply:
Does it mean we are going to call the handler twice in case `if (!APNSRemainedSameDuringFetch && hasFirebaseMessaging)` - one here and one as a result of `[self defaultTokenWithHandler:handler];` in line 915?

Reply:
Good catch. I remove the shouldNotifyHandler as it basically just tells if it's a retry to handle the notify user and handler back.